pkg/controller/useraccount/useraccount_controller.go

```diff
@@ -98,75 +99,74 @@ func (r *ReconcileUserAccount) Reconcile(request reconcile.Request) (reconcile.R
 
 	var created bool
 	var user *userv1.User
-	if user, created, err = r.ensureUser(userAcc); err != nil {
-		_ = r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, err.Error())
+	if user, created, err = r.ensureUser(reqLogger, userAcc); err != nil || created {
 		return reconcile.Result{}, err
-	} else if created {
-		return reconcile.Result{}, r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, "")
 	}
 
 	var identity *userv1.Identity
-	if identity, created, err = r.ensureIdentity(userAcc); err != nil {
-		_ = r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, err.Error())
+	if identity, created, err = r.ensureIdentity(reqLogger, userAcc); err != nil || created {
 		return reconcile.Result{}, err
-	} else if created {
-		return reconcile.Result{}, r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, "")
 	}
 
-	if created, err = r.ensureMapping(userAcc, user, identity); err != nil {
-		_ = r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, err.Error())
+	if created, err = r.ensureMapping(reqLogger, userAcc, user, identity); err != nil || created {
 		return reconcile.Result{}, err
-	} else if created {
-		return reconcile.Result{}, r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, "")
 	}
 
 	return reconcile.Result{}, r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioned, "")
 }
 
-func (r *ReconcileUserAccount) ensureUser(userAcc *toolchainv1alpha1.UserAccount) (*userv1.User, bool, error) {
+func (r *ReconcileUserAccount) ensureUser(logger logr.Logger, userAcc *toolchainv1alpha1.UserAccount) (*userv1.User, bool, error) {
 	user := &userv1.User{}
 	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: userAcc.Name}, user); err != nil {
 		if errors.IsNotFound(err) {
-			log.Info("creating a new user", "name", userAcc.Name)
-			user = newUser(userAcc)
-			if err := controllerutil.SetControllerReference(userAcc, user, r.scheme); err != nil {
+			logger.Info("creating a new user", "name", userAcc.Name)
+			if err := r.updateStatus(userAcc, toolchainv1alpha1.StatusProvisioning, err.Error()); err != nil {
 				return nil, false, err
 			}
-			if err := r.client.Create(context.TODO(), user); err != nil {
-				return nil, false, errs.Wrapf(err, "failed to create user '%s'", userAcc.Name)
+			user = newUser(userAcc)
+			err = controllerutil.SetControllerReference(userAcc, user, r.scheme)
+			if err == nil {
+				err = r.client.Create(context.TODO(), user)
+			}
+			if err != nil {
+				return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, err, "failed to create user '%s'", userAcc.Name)
 			}
-			log.Info("user created successfully", "name", userAcc.Name)
+			logger.Info("user created successfully", "name", userAcc.Name)
 			return user, true, nil
 		}
-		return nil, false, errs.Wrapf(err, "failed to get user '%s'", userAcc.Name)
+		return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, err, "failed to get user '%s'", userAcc.Name)
```

Comment:
In this case, the `wrapErrorWithStatusUpdate` sets the status = `provisioning` and the error to the error msg, but the reconcile doesn't have to be for `POST` request, but for `PATCH`. This means that the status should be rather `updating` instead of `provisioning`. That's why I don't like this automatic setting of the status to `provisioning` when something fails - I would rather use either `failed`,`failing` or `unknown`

Reply:
This is something to discuss...
Let try to look at other examples.

Reply:
according to our discussion either here or in slack channel, it's a bit more complex task and should be handled in separated PR - could you please create a Jira for that?

Reply:
https://jira.coreos.com/browse/CRT-166